libnmstate/nm/context.py

```diff
@@ -62,12 +100,98 @@ def _del_client(self):
                 timeout_source = GLib.timeout_source_new(50)
                 try:
                     timeout_source.set_callback(lambda x: is_timeout.append(1))
-                    timeout_source.attach(self._context)
+                    timeout_source.attach(self.context)
                     while not is_done and not is_timeout:
-                        self._context.iteration(True)
+                        self.context.iteration(True)
                 finally:
                     timeout_source.destroy()
-
             if not is_done:
                 logging.error("BUG: NM.Client is not cleaned")
             self._context = None
+
+    def _del_timeout(self):
+        if self._timeout_source:
+            self._timeout_source.destroy()
+            self._timeout_source = None
+
+    def register_async(self, action, fast=False):
+        """
+        Register action(string) to wait list.
+        """
+        queue = self._fast_queue if fast else self._slow_queue
+        max_queue = FAST_ASYNC_QUEUE_SIZE if fast else SLOW_ASYNC_QUEUE_SIZE
+        if len(queue) >= max_queue:
+            logging.debug(
+                f"Async queue({max_queue}) full, waiting all existing actions "
+                "to be finished before registering more async action"
+            )
+            # TODO: No need to wait all finish, a single finished async is
+            #       sufficient.
+            self.wait_all_finish()
+
+        logging.debug(f"Async action: {action} started")
+        queue.add(action)
```

Comment:
This will also implicitly overwrite an existing action, right? Is this correct? (e.g. registering an action before the previous has been unregistered)
Even if this happens just due to a bug, it may be very hard to track down.

Reply:
Added:

```
        if action in self._fast_queue or action in self._slow_queue:
            raise NmstateInternalError(
                f"BUG: An existing actions {action} is already registered"
            )
```